NullPointerException in CachingTaskQueue

Saw a spate of these, about 300 in 5 minutes. No further instances outside that window.

Version: teku/v0.12.6-dev-5df0d7da/linux-x86_64/-privatebuild-openjdk64bitservervm-java-13

This was the first:
```
05:51:56.813 ERROR - PLEASE FIX OR REPORT | Unexpected exception thrown for beaconchain-async-1
java.util.concurrent.CompletionException: java.lang.NullPointerException
        at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:331) ~[?:?]
        at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:346) ~[?:?]
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:870) ~[?:?]
        at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837) ~[?:?]
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[?:?]
        at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2144) ~[?:?]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$propagateResult$2(SafeFuture.java:131) ~[teku-infrastructure-async-0.12.6-SNAPSHOT.jar:0.12.6-dev-5df0d7da]
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859) ~[?:?]
        at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837) ~[?:?]
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[?:?]
        at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2144) ~[?:?]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$propagateToAsync$9(SafeFuture.java:220) ~[teku-infrastructure-async-0.12.6-SNAPSHOT.jar:0.12.6-dev-5df0d7da]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.of(SafeFuture.java:76) ~[teku-infrastructure-async-0.12.6-SNAPSHOT.jar:0.12.6-dev-5df0d7da]
        at tech.pegasys.teku.infrastructure.async.AsyncRunner.lambda$runAsync$1(AsyncRunner.java:35) ~[teku-infrastructure-async-0.12.6-SNAPSHOT.jar:0.12.6-dev-5df0d7da]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.ofComposed(SafeFuture.java:85) ~[teku-infrastructure-async-0.12.6-SNAPSHOT.jar:0.12.6-dev-5df0d7da]
        at tech.pegasys.teku.infrastructure.async.ScheduledExecutorAsyncRunner.lambda$createRunnableForAction$1(ScheduledExecutorAsyncRunner.java:119) ~[teku-infrastructure-async-0.12.6-SNAPSHOT.jar:>
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:830) [?:?]
Caused by: java.lang.NullPointerException
        at tech.pegasys.teku.core.stategenerator.CachingTaskQueue.completePendingTask(CachingTaskQueue.java:196) ~[teku-ethereum-core-0.12.6-SNAPSHOT.jar:0.12.6-dev-5df0d7da]
        at tech.pegasys.teku.core.stategenerator.CachingTaskQueue.lambda$processNext$6(CachingTaskQueue.java:179) ~[teku-ethereum-core-0.12.6-SNAPSHOT.jar:0.12.6-dev-5df0d7da]
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859) ~[?:?]
        ... 16 more
```

Subsequent ones look like this:

```
05:52:39.029 ERROR - PLEASE FIX OR REPORT | Unexpected exception thrown for beaconchain-async-1
java.util.concurrent.CompletionException: java.lang.NullPointerException
        at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:331) ~[?:?]
        at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:346) ~[?:?]
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:870) ~[?:?]
        at java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:883) ~[?:?]
        at java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2322) ~[?:?]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:389) ~[teku-infrastructure-async-0.12.6-SNAPSHOT.jar:0.12.6-dev-5df0d7da]
        at tech.pegasys.teku.core.stategenerator.CachingTaskQueue.processNext(CachingTaskQueue.java:179) ~[teku-ethereum-core-0.12.6-SNAPSHOT.jar:0.12.6-dev-5df0d7da]
        at tech.pegasys.teku.core.stategenerator.CachingTaskQueue.tryProcessNext(CachingTaskQueue.java:163) ~[teku-ethereum-core-0.12.6-SNAPSHOT.jar:0.12.6-dev-5df0d7da]
        at tech.pegasys.teku.core.stategenerator.CachingTaskQueue.lambda$processNext$7(CachingTaskQueue.java:183) ~[teku-ethereum-core-0.12.6-SNAPSHOT.jar:0.12.6-dev-5df0d7da]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$alwaysRun$16(SafeFuture.java:245) ~[teku-infrastructure-async-0.12.6-SNAPSHOT.jar:0.12.6-dev-5df0d7da]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$exceptionallyCompose$20(SafeFuture.java:293) ~[teku-infrastructure-async-0.12.6-SNAPSHOT.jar:0.12.6-dev-5df0d7da]
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859) ~[?:?]
        at java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:883) ~[?:?]
        at java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2322) ~[?:?]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:389) ~[teku-infrastructure-async-0.12.6-SNAPSHOT.jar:0.12.6-dev-5df0d7da]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.exceptionallyCompose(SafeFuture.java:289) ~[teku-infrastructure-async-0.12.6-SNAPSHOT.jar:0.12.6-dev-5df0d7da]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.alwaysRun(SafeFuture.java:243) ~[teku-infrastructure-async-0.12.6-SNAPSHOT.jar:0.12.6-dev-5df0d7da]
        at tech.pegasys.teku.core.stategenerator.CachingTaskQueue.processNext(CachingTaskQueue.java:180) ~[teku-ethereum-core-0.12.6-SNAPSHOT.jar:0.12.6-dev-5df0d7da]
        at tech.pegasys.teku.core.stategenerator.CachingTaskQueue.tryProcessNext(CachingTaskQueue.java:163) ~[teku-ethereum-core-0.12.6-SNAPSHOT.jar:0.12.6-dev-5df0d7da]
        at tech.pegasys.teku.core.stategenerator.CachingTaskQueue.lambda$processNext$7(CachingTaskQueue.java:183) ~[teku-ethereum-core-0.12.6-SNAPSHOT.jar:0.12.6-dev-5df0d7da]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$alwaysRun$16(SafeFuture.java:245) ~[teku-infrastructure-async-0.12.6-SNAPSHOT.jar:0.12.6-dev-5df0d7da]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$exceptionallyCompose$20(SafeFuture.java:293) ~[teku-infrastructure-async-0.12.6-SNAPSHOT.jar:0.12.6-dev-5df0d7da]
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859) ~[?:?]
        at java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:883) ~[?:?]
        at java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2322) ~[?:?]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:389) ~[teku-infrastructure-async-0.12.6-SNAPSHOT.jar:0.12.6-dev-5df0d7da]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.exceptionallyCompose(SafeFuture.java:289) ~[teku-infrastructure-async-0.12.6-SNAPSHOT.jar:0.12.6-dev-5df0d7da]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.alwaysRun(SafeFuture.java:243) ~[teku-infrastructure-async-0.12.6-SNAPSHOT.jar:0.12.6-dev-5df0d7da]
        at tech.pegasys.teku.core.stategenerator.CachingTaskQueue.processNext(CachingTaskQueue.java:180) ~[teku-ethereum-core-0.12.6-SNAPSHOT.jar:0.12.6-dev-5df0d7da]
        at tech.pegasys.teku.core.stategenerator.CachingTaskQueue.tryProcessNext(CachingTaskQueue.java:163) ~[teku-ethereum-core-0.12.6-SNAPSHOT.jar:0.12.6-dev-5df0d7da]
        at tech.pegasys.teku.core.stategenerator.CachingTaskQueue.lambda$processNext$7(CachingTaskQueue.java:183) ~[teku-ethereum-core-0.12.6-SNAPSHOT.jar:0.12.6-dev-5df0d7da]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$alwaysRun$16(SafeFuture.java:245) ~[teku-infrastructure-async-0.12.6-SNAPSHOT.jar:0.12.6-dev-5df0d7da]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$exceptionallyCompose$20(SafeFuture.java:293) ~[teku-infrastructure-async-0.12.6-SNAPSHOT.jar:0.12.6-dev-5df0d7da]
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859) ~[?:?]
        at java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:883) ~[?:?]
        at java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2322) ~[?:?]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:389) ~[teku-infrastructure-async-0.12.6-SNAPSHOT.jar:0.12.6-dev-5df0d7da]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.exceptionallyCompose(SafeFuture.java:289) ~[teku-infrastructure-async-0.12.6-SNAPSHOT.jar:0.12.6-dev-5df0d7da]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.alwaysRun(SafeFuture.java:243) ~[teku-infrastructure-async-0.12.6-SNAPSHOT.jar:0.12.6-dev-5df0d7da]
        at tech.pegasys.teku.core.stategenerator.CachingTaskQueue.processNext(CachingTaskQueue.java:180) ~[teku-ethereum-core-0.12.6-SNAPSHOT.jar:0.12.6-dev-5df0d7da]
        at tech.pegasys.teku.core.stategenerator.CachingTaskQueue.tryProcessNext(CachingTaskQueue.java:163) ~[teku-ethereum-core-0.12.6-SNAPSHOT.jar:0.12.6-dev-5df0d7da]
        at tech.pegasys.teku.core.stategenerator.CachingTaskQueue.lambda$processNext$7(CachingTaskQueue.java:183) ~[teku-ethereum-core-0.12.6-SNAPSHOT.jar:0.12.6-dev-5df0d7da]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$alwaysRun$16(SafeFuture.java:245) ~[teku-infrastructure-async-0.12.6-SNAPSHOT.jar:0.12.6-dev-5df0d7da]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$exceptionallyCompose$20(SafeFuture.java:293) ~[teku-infrastructure-async-0.12.6-SNAPSHOT.jar:0.12.6-dev-5df0d7da]

<...repeats the identical block of stack trace about 100 times>

Caused by: java.lang.NullPointerException
```

--- a/ethereum/core/src/main/java/tech/pegasys/teku/core/stategenerator/CachingTaskQueue.java
+++ b/ethereum/core/src/main/java/tech/pegasys/teku/core/stategenerator/CachingTaskQueue.java
@@ -141,7 +141,7 @@ public class CachingTaskQueue<K, V> {
       newBase
           .get()
           .thenAccept(ancestorResult -> queueTask(task.rebase(ancestorResult)))
-          .propagateExceptionTo(generationResult);
+          .finish(error -> completePendingTask(task, Optional.empty(), error));
       return generationResult;
     }
 

--- a/ethereum/core/src/main/java/tech/pegasys/teku/core/stategenerator/CachingTaskQueue.java
+++ b/ethereum/core/src/main/java/tech/pegasys/teku/core/stategenerator/CachingTaskQueue.java
@@ -123,6 +123,9 @@ public class CachingTaskQueue<K, V> {
       return currentPendingTask;
     }
 
+    final SafeFuture<Optional<V>> generationResult = new SafeFuture<>();
+    pendingTasks.put(task.getKey(), generationResult);
+
     // Check if there's a better starting point (in cache or in progress)
     final Optional<SafeFuture<Optional<V>>> newBase =
         task.streamIntermediateSteps()
@@ -135,24 +138,26 @@ public class CachingTaskQueue<K, V> {
             .findFirst();
     if (newBase.isPresent()) {
       rebasedTaskCounter.inc();
-      return newBase.get().thenCompose(ancestorResult -> queueTask(task.rebase(ancestorResult)));
+      newBase
+          .get()
+          .thenAccept(ancestorResult -> queueTask(task.rebase(ancestorResult)))
+          .propagateExceptionTo(generationResult);
+      return generationResult;
     }
 
     // Schedule the task for execution
     newTaskCounter.inc();
-    return queueTask(task);
+    queueTask(task);
+    return generationResult;
   }
 
   public Optional<V> getIfAvailable(final K key) {
     return Optional.ofNullable(cache.get(key));
   }
 
-  private SafeFuture<Optional<V>> queueTask(final CacheableTask<K, V> task) {
-    final SafeFuture<Optional<V>> generationResult = new SafeFuture<>();
-    pendingTasks.put(task.getKey(), generationResult);
+  private void queueTask(final CacheableTask<K, V> task) {
     queuedTasks.add(task);
     tryProcessNext();
-    return generationResult;
   }
 
   private synchronized void tryProcessNext() {

--- a/ethereum/core/src/test/java/tech/pegasys/teku/core/stategenerator/CachingTaskQueueTest.java
+++ b/ethereum/core/src/test/java/tech/pegasys/teku/core/stategenerator/CachingTaskQueueTest.java
@@ -154,6 +154,8 @@ class CachingTaskQueueTest {
     final SafeFuture<Optional<String>> resultB = taskQueue.perform(taskB);
     final SafeFuture<Optional<String>> resultC = taskQueue.perform(taskC);
 
+    assertPendingTaskCount(2); // B & C were de-duplicated
+
     taskA.assertPerformedWithoutRebase();
     // Task B will be scheduled for rebase when A completes
     // Task C should just use the pending future from B and never execute
@@ -167,6 +169,7 @@ class CachingTaskQueueTest {
 
     taskB.completeTask();
 
+    assertThat(resultA).isCompletedWithValue(taskA.getExpectedValue());
     assertThat(resultB).isCompletedWithValue(taskB.getExpectedValue());
     assertThat(resultC).isCompletedWithValue(taskC.getExpectedValue());
     taskC.assertNotRebased();
@@ -269,6 +272,14 @@ class CachingTaskQueueTest {
         metricsSystem
             .getCounter(TekuMetricCategory.STORAGE, METRICS_PREFIX + "_tasks_total")
             .getValue("rebase");
+    assertThat(value).isEqualTo(expectedCount);
+  }
+
+  private void assertPendingTaskCount(final int expectedCount) {
+    final double value =
+        metricsSystem
+            .getGauge(TekuMetricCategory.STORAGE, METRICS_PREFIX + "_tasks_requested")
+            .getValue();
     assertThat(value).isEqualTo(expectedCount);
   }
 

--- a/ethereum/core/src/test/java/tech/pegasys/teku/core/stategenerator/CachingTaskQueueTest.java
+++ b/ethereum/core/src/test/java/tech/pegasys/teku/core/stategenerator/CachingTaskQueueTest.java
@@ -145,6 +145,35 @@ class CachingTaskQueueTest {
   }
 
   @Test
+  void shouldNotPerformDuplicateTasksWhenTasksAreRebased() {
+    final StubTask taskA = new StubTask(1);
+    final StubTask taskB = new StubTask(5, 1);
+    final StubTask taskC = new StubTask(5, 1);
+
+    final SafeFuture<Optional<String>> resultA = taskQueue.perform(taskA);
+    final SafeFuture<Optional<String>> resultB = taskQueue.perform(taskB);
+    final SafeFuture<Optional<String>> resultC = taskQueue.perform(taskC);
+
+    taskA.assertPerformedWithoutRebase();
+    // Task B will be scheduled for rebase when A completes
+    // Task C should just use the pending future from B and never execute
+    taskC.assertNotRebased();
+    taskC.assertNotPerformed();
+
+    taskA.completeTask();
+    taskB.assertPerformedFrom(taskA.getExpectedValue().orElseThrow());
+    taskC.assertNotRebased();
+    taskC.assertNotPerformed();
+
+    taskB.completeTask();
+
+    assertThat(resultB).isCompletedWithValue(taskB.getExpectedValue());
+    assertThat(resultC).isCompletedWithValue(taskC.getExpectedValue());
+    taskC.assertNotRebased();
+    taskC.assertNotPerformed();
+  }
+
+  @Test
   void getIfAvailable_shouldReturnValueWhenPresent() {
     final StubTask task = new StubTask(1);
     final SafeFuture<Optional<String>> result = taskQueue.perform(task);

